File: frontend/src/api/subjects.ts
import axios from "axios";

const API_URL = "http://127.0.0.1:8000";


export interface AvailableSubject {
  subject_code: string;
  subject_title: string;
}


interface AvailableSubjectsResponse {
  message: string;
  data: AvailableSubject[];
}


export async function getAvailableSubjects():
  Promise<AvailableSubject[]> {

  const response =
    await axios.get<AvailableSubjectsResponse>(
      `${API_URL}/api/subjects/available`
    );

  return response.data.data;
}

File: backend/routers/subjects.py
from fastapi import APIRouter

from genetic_algorithm.utils.Functions import list_subjects


router = APIRouter(
    prefix="/api/subjects",
    tags=["Subjects"],
)


@router.get("/available")
def get_available_subjects():

    unique_subjects = {}

    for subject in list_subjects:

        code = str(
            getattr(
                subject,
                "number",
                ""
            )
        ).strip().upper()

        title = str(
            getattr(
                subject,
                "title",
                ""
            )
            or ""
        ).strip()

        if not code:
            continue

        if code not in unique_subjects:

            unique_subjects[code] = {
                "subject_code": code,
                "subject_title": title,
            }

    subjects = sorted(
        unique_subjects.values(),
        key=lambda item:
            item["subject_code"]
    )

    return {
        "message":
            "Available subjects retrieved successfully.",

        "data":
            subjects,
    }